### PR TITLE
Unroll loop support for free0, free1 dimension

### PR DESCRIPTION
**Pull request by ramjana**
-support to unroll in Free0,Free1 dimension.

### CODE REVIEW

**Comment by zaliu** [comment]
@ramjana Please resolve conflicts so that CI can run.  Thanks.

File: Tensile/KernelWriter.py [hunk 1]
```diff
@@ -836,49 +1044,76 @@ def noLoadLoop( self, kernel, tensorParametersA, tensorParametersB, isOptNLL, pa
...
+                  pack[plrPIdx].addCode(packCodeB)
+                  #if (kernel["NumITilesUnRoll"] + kernel["NumJTilesUnRoll"] > 2 and self.scheduleIterAlg ==2): 
+                  localReads.addText(self.comment("local read inc a"))
+                  localReads.addText(self.localReadInc(kernel, iui, tensorParametersA))
+                  localReads.addText(self.comment("local read inc b"))
+                  localReads.addText(self.localReadInc(kernel, iui, tensorParametersB))
+          # use only when unorllI,unrollJ != 1
+          if not (self.scheduleLastIterAlg):  
+            if localReads:
+              k1.append(localReads)
```
Comment (AlexBrownAMD) [inline Tensile/KernelWriter.py:1081]: This section could be restructured: if using scheduleLastIterAlg, the loop doesn't need to run at all, and putting the check before the loop would also make it easier to read.

File: Tensile/KernelWriter.py [hunk 2]
```diff
@@ -836,49 +1044,76 @@ def noLoadLoop( self, kernel, tensorParametersA, tensorParametersB, isOptNLL, pa
...
+            if localReads:
+              k1.append(localReads) 
+          if self.enable["Wait"]:
+            dataAtIter = u//self.numIterPerCoalescedRead - self.numItersPLR
+            numReadsIter = min(u+1,kernel["LoopIters"]//self.numIterPerCoalescedRead - self.numItersPLR)
+            skipReadsIter = numReadsIter - dataAtIter - 1
+            waitCode = self.wait(kernel, tensorParametersA, tensorParametersB, -1, -1, \
+                skipReadsIter, \
+                "7wait for local read")
+          if not (self.scheduleLastIterAlg):
```
Comment (AlexBrownAMD) [inline Tensile/KernelWriter.py:1089]: There's a number of these checks in this section of code, maybe a refactor like using modules could help with ease of reading in this section.

File: Tensile/KernelWriter.py [hunk 3]
```diff
@@ -309,6 +309,216 @@ def makeSchedule(self, kernel, tensorParametersA, tensorParametersB, localWriteE
...
+          packA = Code.Module()
+          packB = Code.Module()
+          packToSched = False
+          movaccVgprToSched = False
+          storeCToSched = False
+
+          if packAB.items():
+             packA.addCode(packAB.items().pop(0))
+             packInstPerMfma = packA.count()
+          if packAB.items():
```
Comment (AlexBrownAMD) [inline Tensile/KernelWriter.py:460]: Same if condition as above

File: Tensile/KernelWriter.py [hunk 4]
```diff
@@ -309,6 +309,216 @@ def makeSchedule(self, kernel, tensorParametersA, tensorParametersB, localWriteE
       # should never run out of items to schedule
       assert not itemsToSched # should have scheduled everthing already
 
+  def makeLastIterSchedule(self, kernel, localReadCode, waitCode, macIterCode, \
+                                 packCode = Code.Module(), movAccVgprCodeItems = Code.Module() ):
+
+    iterCode = Code.Module()
+
+    # Default schedule is other, waitCode, MAC/MFMA code:
+    if self.scheduleLastIterAlg == 1:
```
Comment (AlexBrownAMD) [inline Tensile/KernelWriter.py:318]: This should probably go in a module

File: Tensile/KernelWriter.py [hunk 5]
```diff
@@ -309,6 +309,216 @@ def makeSchedule(self, kernel, tensorParametersA, tensorParametersB, localWriteE
...
+              availableCyclesforRead = latencyPerMfmaIter - (packInstPerMfma*4) - 8
+            else: 
+              availableCyclesforRead = latencyPerMfmaIter - 8
+            numReadsToSched = availableCyclesforRead // 8
+            if packToSched:
+              numReadsToSched = numReadsToSched - packInstPerMfma*4
+            if (i == 0):
+              numReadsToSched  = 4 if (numReadsToSched>4) else numReadsToSched  # cap it to 4 for credit-max 
+            else:
+              numReadsToSched  = 3 if (numReadsToSched>3) else numReadsToSched  # cap it to 3  to allow other instructions
```
Comment (AlexBrownAMD) [inline Tensile/KernelWriter.py:483]: should if (i==0) be elif? Can this block be combined with the similar if/else block above it?

**Review by AlexBrownAMD — CHANGES_REQUESTED** [review]

**Review by cgmillette — CHANGES_REQUESTED** [review]
Builds are failing, plus general code structure comments

**Comment by zaliu** [comment]
Inactivity.  Can reopen later.

**Pull request closed without merging**